### ISSUE TITLE
beginShape() endShape(CLOSE): limited to 3 vertices

### ISSUE THREAD

**Issue by whoem** [processing/p5.js#1209]
When I want to draw a shape using beginshape() endShape(CLOSE) with more than 3 vertices, it will only use the first 3 vertices in stack and form a triangle. Example:
```
function setup() {
  createCanvas(500, 500, WEBGL);
}

function draw() {
  background(0);
  fill (255);

  beginShape();
  vertex (-100, -100, 100);
  vertex (100, -100, 100);
  vertex (100, 100, 100);
  vertex (-100, 100, 100);
  endShape (CLOSE);
}
```
... creates the following image:
![triangle](https://cloud.githubusercontent.com/assets/6699384/12366264/5b1a079c-bbda-11e5-8a29-d26298adaa9f.png)

:(

**Comment by parsoyaarihant** [comment]
hi, this issue is fixed in the webgl-geometry-refractor branch.

**Comment by whoem** [comment]
Hi, with webgl-geometry-refractor branch I get not even a white triangle with the same sketch. Just a black background with nothing in it.

**Comment by parsoyaarihant** [comment]
this bug was fixed once by @indefinit  but other problems emerged again so u can try parsoyaarihant:implement-bezier- branch to run this but there may be other problems such as light properties. Refer to PR #1204.

**Comment by whoem** [comment]
Right, it works. Awesome work! :dancer: When will this be merged into master branch?

**Comment by indefinit** [comment]
@parsoyaarihant I've tested this on my system and I'm seeing correct behavior from beginShape() endShape() from the geometry-refactor branch.  This means you should be able to call lights and textures in addition to fill() on beginShape() without any errors.

@whoem we're getting close to merge time.  The items we need to finish before merging are correctly working cylinder and cone geometries as well as the completion of the .obj loader, which makes full use of the refactored p5.Geometry object.

**Comment by whoem** [comment]
@indefinit Indeed in the most recent geometry-refractor branch, beginShape() endShape() works neat and calling lights works basically. But lights seem to address primitives only and ignore beginShape() endShape() elements.
Also, if you have a construction like this:
```
draw() {
  anyLight (direction, position);
  rotateX (angle);
  rotateY (angle);
  object (position);
}
```
... eventhough you call the light before the transformations, they will take effect on it. Strange.

Another thing: The point of origin in WEBGL mode is set at the center of the canvas. In p5 it's still at the top-left corner of the canvas. Is this on purpose?

I am very happy to hear, that you are about to be done refractoring these heaps of lines of code :) Never easy. Never joyful.

**Comment by karenpeng** [comment]
@indefinit @parsoyaarihant thanks for your works on immediate mode!
@whoem the origin point in WebGL mode is in the center. Where did you see that's in the top-left corner, is it after translate?  And about the lighting, currently there's a problem in the world position,  #1194 is my thought to tackle it, in which the world position won't be effected by the transformation.

**Comment by whoem** [comment]
@karenpeng I was comparing WebGL's behaviour to OpenGL's origin point handling on desktop java p5. Obviously not the same, but do you intend to do so? I mean, is it a design aim of p5.js to imitate desktop p5's behaviour or "only" to translate it's methodology to web standards? I don't judge, just wanna know. Subscribed #1194. Thanks for the clue.

**Comment by karenpeng** [comment]
@whoem oh yes you are right! It's an intended adaptation from java processing. We made this changed mainly because, if you look at cg fundamental and 3d softwares, center origin is widely used. There's a camera in 3d, it determines how you see the world. Even though the origin of the world is set to be the top left corner, it does not necessarily put the shape in the exact top left corner because of the camera.
But if the origin is in the center, the shape will always appear in the center, this won't cause any confusions to users.
Hopefully this [get started](https://github.com/processing/p5.js/wiki/Getting-started-with-WebGL-in-p5) tutorial will clear more confusions.

**Comment by whoem** [comment]
@karenpeng Thanks. Already knew the tutorial, but just focused on the beginShape() endShape() part ;)

**Comment by indefinit** [comment]
Closing this Issue because it is fixed in the geometry-refactor branch.  Will be merged into master shortly.
